Enhanced Pod Failure Logging in KubernetesExecutor

HsiuChuanHsu opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->

## Description
Add detailed logging of pod and container failure information when tasks fail, including pod phase, reasons, messages, and container states for better debugging.

## Problem Statement
When tasks fail in KubernetesExecutor, operators often lack sufficient information to quickly diagnose the root cause of failures. The current implementation provides minimal failure context, making troubleshooting time-consuming and inefficient.

## Solution
Introduced comprehensive logging for pod and container failure states, capturing detailed error information to streamline debugging.

## Changes Made
Core Implementation
- Enhanced `_change_state` method in `KubernetesExecutor`
- Added pod status extraction using `kube_client.read_namespaced_pod()`
- Implemented container state analysis for terminated and waiting states
- Added graceful exception handling for Kubernetes API failures

Closes: #37548 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

HsiuChuanHsu committed:
feat: Enhanced pod failure logging in KubernetesExecutor

Add detailed logging of pod and container failure information when tasks fail,
including pod phase, reasons, messages, and container states for better debugging.

- Extract pod status (phase, reason, message) on task failure
- Extract container state info (terminated/waiting reasons & messages)
- Add exception handling for Kubernetes API failures
- Only execute additional logging for FAILED task states

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+                    pod_message = getattr(pod.status, "message", None)
+
+                    # If containerStatuses has detailed reasons, print them as well.
+                    container_statuses = getattr(pod.status, "container_statuses", None)
```
Note that there is also a `pod.status.init_container_statuses` attribute, it may be good to include those errors as well.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
 
+        if state == TaskInstanceState.FAILED:
+            try:
+                if self.kube_client:
```
In what scenario would `self.kube_client` not be set here?

HsiuChuanHsu replied:
After moving this logic to `KubernetesJobWatcher.process_status`, we don't have to worry about `self.kube_client` could be `None`.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+                                container_message = getattr(state_obj.waiting, "message", None)
+                                break
+                    self.log.error(
+                        "Pod %s in namespace %s failed. Pod phase: %s, reason: %s, message: %s, container_state: %s, container_reason: %s, container_message: %s",
```
Please include `key` here. Or at least the `run_id` component of the `key`. It makes it a lot easier to search through logs without having to consider previous lines "recently logged".

HsiuChuanHsu replied:
Yeah!! This is also a good part to add `key` into log. I make the key info into this format `task_key_str = f"{key.dag_id}.{key.task_id}.{key.try_number}"`.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+        if state == TaskInstanceState.FAILED:
+            try:
+                if self.kube_client:
+                    pod = self.kube_client.read_namespaced_pod(name=pod_name, namespace=namespace)
```
Would it be better to do all this inside class `KubernetesJobWatcher.process_status`? That's where the pod status is being read today, to me it feels like that would be more consistent, so that `change_state` and `process_status` does not have different views of what is a failure and not, and it reduces the number of API calls to k8s. 

I'm not 100% sure but i also believe that `process_status` is being run on a separate thread, communicating to the other thread through the `result_queue`, if k8s API calls are being done on the result-thread it may introduce unexpected latencies there.

See for example https://github.com/apache/airflow/pull/39924 where similar debug improvements were done in `process_status`.

Would be good if someone more familiar with the architecture of this class could chime in here.

HsiuChuanHsu replied:
True!!! Moving this logic to `KubernetesJobWatcher.process_status` would be much consistent, since `process_status` already reads pod status to determine task state. 

This change would reduces load on Kubernetes API server and no need for and additional `read_namespaced_pod` call in `_change_state`.

Thanks for this suggestion!

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+                            state_obj = cs.state
+                            if state_obj.terminated:
+                                container_state = "terminated"
+                                container_reason = getattr(state_obj.terminated, "reason", None)
```
When `reason == "Completed"` it means the pod ran to completion as expected. It should perhaps not be logged as error,  Though the `state` would most likely not be `FAILED` in that case. Only if there are multiple `container_statuses`, and since there is a for-loop over them it may appear that that scenario is supported.

HsiuChuanHsu replied:
True! 
Fix: Only containers that actually failed (non-zero exit code AND reason != "Completed") are reported as failures.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+                                container_reason = getattr(state_obj.waiting, "reason", None)
+                                container_message = getattr(state_obj.waiting, "message", None)
+                                break
+                    self.log.error(
```
Should this really be an error from scheduler point of view? If a DAG failed, the scheduler has still behaved as expected. Maybe info or warning?

(I would also love to see these errors surfaced to the log in the UI but as far as i know the architecture doesn't support that yet, it's only scheduler administrators that can access them now. This is still a good improvement as is :) )

HsiuChuanHsu replied:
I temporarily make the log as warning. Yes, from the scheduler's perspective  it's just normal business logic where a task didn't succeed.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py:
```diff
@@ -417,6 +417,46 @@ def _change_state(
+                        container_message,
+                    )
+            except Exception as e:
+                self.log.warning("Failed to fetch pod failure reason for %s/%s: %s", namespace, pod_name, e)
```
Same here, please include `key` in log entry.

HsiuChuanHsu replied:
Added!

hterik reviewed (commented):
Nice! :+1:

HsiuChuanHsu committed:
feat(k8s-executor): improve pod failure diagnostics and reduce API calls

Move failure analysis to watcher thread, add detailed container status logging,
and include task keys for better log searchability.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_types.py:
```diff
@@ -31,11 +31,15 @@
+    # pod_name, namespace, pod state, annotations, resource_version, failure_details
+    KubernetesWatchType = tuple[
+        str, str, TaskInstanceState | str | None, dict[str, str], str, dict[str, Any] | None
+    ]
```
Prefer `typing.NamedTuple`. Tuples with this many entries are very difficult to reason about otherwise.

HsiuChuanHsu replied:
Thank you for the suggestion! 
I agree that using `typing.NamedTuple` would improve readability for tuples with many entries. However, I'd prefer to address this in a separate PR. This refactoring would likely require updates to many other functions that consume these tuples.

hterik replied:
Sounds good :+1:

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_types.py:
```diff
@@ -31,11 +31,15 @@
-    KubernetesWatchType = tuple[str, str, TaskInstanceState | str | None, dict[str, str], str]
+    # key, pod state, pod_name, namespace, resource_version, failure_details
+    KubernetesResultsType = tuple[
+        TaskInstanceKey, TaskInstanceState | str | None, str, str, str, dict[str, Any] | None
```
Prefer a `TypedDict` or `dataclass` for the Failure details. Untyped `dicts` are very easy to get wrong due to typos etc.

Maybe such FailureDetails class even deserves it's own custom `__str__` implementation, to avoid all the `.get` in `_change_state`?

HsiuChuanHsu replied:
Agree! I've switched it to use a TypedDict for the failure details now. Much safer than plain dicts.

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -210,22 +210,115 @@ def process_status(
+            except Exception as e:
+                self.log.warning(
+                    "Failed to collect pod failure details for %s/%s: %s", namespace, pod_name, e
+                )
```
Suggest refactoring this into a function: `def collect_failures(pod: V1Pod) -> FailureDetails`. With `FailureDetails` being `TypedDict` as suggested above.

HsiuChuanHsu replied:
Great idea! Already done - I've extracted this into a `collect_pod_failure_details` function that returns FailureDetails (TypedDict). Much cleaner now, thanks!

hterik reviewed (commented):
Some minor improvements suggested. Otherwise looks ok to me. 

Someone else with more knowledge and authority should take a look as well.

HsiuChuanHsu committed:
refactor: Extract pod failure analysis to dedicated TypedDict functions

- Add FailureDetails TypedDict for type-safe failure information
- Extract collect_pod_failure_details() function for cleaner separation of concerns
- Create _analyze_init_containers() and _analyze_main_containers() helper functions

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -302,6 +353,139 @@ def process_status(
+
+    except Exception:
+        # Return basic pod info if container analysis fails
+        return {
```
Please log this exception before returning, using `logger.exception`. It's an unexpected scenario and if it happens, then administrators should have the info required to debug it.

hterik replied:
Done

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -302,6 +353,139 @@ def process_status(
+                    "container_type": "main",
+                    "container_name": getattr(cs, "name", "unknown"),
+                },
+            )
```
I believe this part is exactly the same as in `_analyze_init_containers`. Does it make sense to reuse?

hterik reviewed (commented):
LGTM! :+1:

HsiuChuanHsu committed:
feat(kubernetes): enhance pod failure diagnostics with TypedDict and modular design
1. Introduce FailureDetails TypedDict
2. Implement collect_pod_failure_details Function
3. Enhanced Failure Logging

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -302,6 +353,114 @@ def process_status(
+        }
+
+
+def _analyze_containers(container_statuses: list | None, container_type: str) -> FailureDetails | None:
```
list[k8s.V1Container]

HsiuChuanHsu replied:
True! Make it more safer.
`def _analyze_containers(
    container_statuses: list[k8s.V1ContainerStatus] | None, container_type: Literal["init", "main"]
) `

hterik commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_types.py:
```diff
@@ -16,9 +16,25 @@
+    container_reason: str | None
+    container_message: str | None
+    exit_code: int | None
+    container_type: str | None  # "init" or "main"
```
`Literal["init","main"] | None`

HsiuChuanHsu replied:
Fixed !

hterik reviewed (commented):
Minor nitpicks from me only, otherwise looks good. 
Someone else need to approve.

HsiuChuanHsu committed:
fix(kubernetes): improve type safety with Literal types for container analysis

jason810496 commented on providers/cncf/kubernetes/tests/unit/cncf/kubernetes/executors/test_kubernetes_executor.py:
```diff
@@ -53,7 +53,7 @@
 )
 from airflow.providers.standard.operators.empty import EmptyOperator
-from airflow.utils import timezone
+from airflow.utils import timezone  # type: ignore[attr-defined]
```
The `timezone` has been moved to new location.

```suggestion
try:
    from airflow.sdk import timezone
except ImportError:
    from airflow.utils import timezone  # type: ignore[attr-defined,no-redef]
```

HsiuChuanHsu replied:
Fixed! Thanks for reminder.

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -210,22 +211,42 @@ def process_status(
+                    TaskInstanceState.FAILED,
+                    annotations,
+                    resource_version,
+                    failure_details,
```
```suggestion
                    None,
```

If I don't get it wrong, it seems that only `status == "Failed"` case will call `failure_details = collect_pod_failure_details(pod)`.

HsiuChuanHsu replied:
Fixed!

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -254,28 +275,51 @@ def process_status(
+                                    TaskInstanceState.FAILED,
+                                    annotations,
+                                    resource_version,
+                                    failure_details,
```
```suggestion
                                    None,
```

HsiuChuanHsu replied:
Fixed!

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -254,28 +275,51 @@ def process_status(
+            task_key_str = f"{key.dag_id}.{key.task_id}.{key.try_number}" if key else "unknown"
+            self.log.warning(
+                "Event: %s Failed, task: %s, annotations: %s", pod_name, task_key_str, annotations_string
+            )
```
Would it be better to call `collect_pod_failure_details` only within the `status == "Failed"` block? For other cases, `failure_details` should remain `None`.

HsiuChuanHsu replied:
Fixed this part! True, just keep the failure_details from `status == "Failed"`

jason810496 commented on providers/cncf/kubernetes/src/airflow/providers/cncf/kubernetes/executors/kubernetes_executor_utils.py:
```diff
@@ -302,6 +353,116 @@ def process_status(
+        logging.getLogger(__name__).exception(
+            "Unexpected error while collecting pod failure details for pod %s",
+            getattr(pod.metadata, "name", "unknown"),
+        )
```
How about passing the logger from `KubernetesJobWatcher.log` property as parameter instead of creating a new one?

HsiuChuanHsu replied:
Fixed!

jason810496 reviewed (commented):
Nice! Thanks for the PR!
Only a question regarding where to call `collect_pod_failure_details(pod)`.

HsiuChuanHsu committed:
feat: enhance pod failure logging in KubernetesExecutor with detailed failure analysis
- Pass logger as parameter to collect_pod_failure_details for consistent logging context

- Move failure details collection logic to only execute for Failed status pods for better performance
- Update test imports to handle timezone module location changes between Airflow versions
